psydac/api/discretization.py

```diff
@@ -270,192 +268,91 @@ def solve(self, **kwargs):
-            d_spaces[1].diff = d_spaces[1].curl = D1
-            
+#==============================================================================
+def reduce_space_degrees(V, Vh, basis='B'):
+
+    if isinstance(V.kind, HcurlSpaceType):
+        if V.ldim == 2:
+            spaces = [Vh.reduce_degree(axes=[0], basis=basis),
+                      Vh.reduce_degree(axes=[1], basis=basis)]
+        elif V.ldim == 3:
+            spaces = [Vh.reduce_degree(axes=[0], basis=basis),
+                      Vh.reduce_degree(axes=[1], basis=basis),
+                      Vh.reduce_degree(axes=[2], basis=basis)]
         else:
-            D0 = VectorCurl_2D(d_spaces[0], d_spaces[1])
-            D1 = Divergence_2D(d_spaces[1], d_spaces[2])
-
-            d_spaces[0].diff = d_spaces[0].rot = D0
-            d_spaces[1].diff = d_spaces[1].div = D1
+            raise NotImplementedError('TODO')
+
+        Vh = ProductFemSpace(*spaces)
+    elif isinstance(V.kind, HdivSpaceType):
+
+        if V.ldim == 2:
+            spaces = [Vh.reduce_degree(axes=[1], basis=basis),
+                      Vh.reduce_degree(axes=[0], basis=basis)]
+        elif V.ldim == 3:
+            spaces = [Vh.reduce_degree(axes=[1,2], basis=basis),
+                      Vh.reduce_degree(axes=[0,2], basis=basis),
+                      Vh.reduce_degree(axes=[0,1], basis=basis)]
+        else:
+            raise NotImplementedError('TODO')
 
-    elif ldim == 3:
+        Vh = ProductFemSpace(*spaces)
 
-        d_spaces[0] = discretize_space(spaces[0], domain_h, *args, basis='B', **kwargs)
-        d_spaces[1] = discretize_space(spaces[1], domain_h, *args, basis='M', **kwargs)
-        d_spaces[2] = discretize_space(spaces[2], domain_h, *args, basis='M', **kwargs)
-        d_spaces[3] = discretize_space(spaces[3], domain_h, *args, basis='M', **kwargs)
+    elif isinstance(V.kind, L2SpaceType):
+        if V.ldim == 1:
+            Vh = Vh.reduce_degree(axes=[0], basis=basis)
+        elif V.ldim == 2:
+            Vh = Vh.reduce_degree(axes=[0,1], basis=basis)
+        elif V.ldim == 3:
+            Vh = Vh.reduce_degree(axes=[0,1,2], basis=basis)
 
-        D0 =   Gradient_3D(d_spaces[0], d_spaces[1])
-        D1 =       Curl_3D(d_spaces[1], d_spaces[2])  
-        D2 = Divergence_3D(d_spaces[2], d_spaces[3])
+    elif not isinstance(V.kind,  (H1SpaceType, UndefinedSpaceType)):
+        raise NotImplementedError('TODO')
 
-        d_spaces[0].diff = d_spaces[0].grad = D0
-        d_spaces[1].diff = d_spaces[1].curl = D1
-        d_spaces[2].diff = d_spaces[2].div  = D2
+    if isinstance(V, VectorFunctionSpace):
+        if isinstance(V.kind, (H1SpaceType, L2SpaceType, UndefinedSpaceType)):
+            Vh = ProductFemSpace(*[Vh]*V.ldim)
 
-    return DiscreteDerham(mapping, *d_spaces)
+    return Vh
 
 #==============================================================================
-# TODO multi patch
 # TODO knots
 def discretize_space(V, domain_h, *args, **kwargs):
```

Comment:
Can you explain a bit how the algorithm that is used in this function? It is hard to follow without any comment in the code.

Reply:
okay, the discretize_space is divided into two cases, the case where we have a geometry file and the case where we have either an analytical mapping or our domain is the logical domain.
In the two cases, we build the dictionary `g_spaces` that contains the interiors as keys and the spaces as value, In the case of single patch , we have one key and one value

Reply:
after building the `g_spaces` we create the compatible spaces if needed in each interior domain

Reply:
Thanks! Can you add a nice docstring to the function, with this explanation for future reference?

Reply:
okay

Reply:
done